File: solutions/python/matching-brackets/1/matching_brackets.py
mem = []


def is_paired(input_string):
    for char in input_string:
        if char in '[({':
            mem.extend(char)
        elif char in '])}':
            try:
                last = mem.pop()
                if not is_opposite(last, char):
                    return False
            except IndexError:
                return False
    return len(mem) == 0


def is_opposite(saved: str, current: str) -> bool:
    match current:
        case ']':
            return saved == '['
        case '}':
            return saved == '{'
        case ')':
            return saved == '('
        case _:
            return False


# if __name__ == '__main__':
#     # print(is_paired("{}[]"))
#     print(is_paired("{[]}"))

File: solutions/typescript/matching-brackets/1/matching-brackets.ts
const braces = new Map([
  ["]", "["],
  ["}", "{"],
  [")", "("]
]);

export function isPaired(input: string): boolean {
  const mem: string[] = []

  for (let bracket of input.split("")) {
    if ("[{(".indexOf(bracket) > -1) {
      mem.push(bracket)
    } else if ("]})".indexOf(bracket) > -1) {
      const closing = braces.get(bracket)
      if (closing) {
        if (closing !== mem.pop()) {
          return false
        }
      }
    }
  }

  return mem.length === 0
}